packages/next-upgrade/index.ts

```diff
@@ -0,0 +1,223 @@
+#!/usr/bin/env node
+
+import prompts from 'prompts'
+import fs from 'fs'
+import { execSync } from 'child_process'
+import path from 'path'
+import { compareVersions } from 'compare-versions'
+import yaml from 'yaml'
+
+type VersionSpecifier = 'canary' | 'rc' | 'latest'
+type PackageManager = 'pnpm' | 'npm' | 'yarn'
+
+interface Response {
+  version: VersionSpecifier
+}
+
+async function run(): Promise<void> {
+  let targetVersionSpecifier: VersionSpecifier
+
+  let nextPackageJson: { [key: string]: any } = {}
+  try {
+    const resCanary = await fetch(`https://registry.npmjs.org/next/canary`)
+    nextPackageJson['canary'] = await resCanary.json()
+
+    const resRc = await fetch(`https://registry.npmjs.org/next/rc`)
+    nextPackageJson['rc'] = await resRc.json()
+
+    const resLatest = await fetch(`https://registry.npmjs.org/next/latest`)
+    nextPackageJson['latest'] = await resLatest.json()
+  } catch (error) {
+    console.error('Failed to fetch versions from npm registry.')
+    return
+  }
+
+  let showRc = true
+  if (nextPackageJson['latest'].version && nextPackageJson['rc'].version) {
+    showRc =
+      compareVersions(
+        nextPackageJson['rc'].version,
+        nextPackageJson['latest'].version
+      ) === 1
+  }
+
+  const choices = [
+    {
+      title: 'Canary',
+      value: 'canary',
+      description: `Experimental version with latest features (${nextPackageJson['canary'].version})`,
+    },
+  ]
+  if (showRc) {
+    choices.push({
+      title: 'Latest Release Candidate',
+      value: 'rc',
+      description: `Pre-release version for final testing (${nextPackageJson['rc'].version})`,
+    })
+  }
+  choices.push({
+    title: 'Latest Stable',
+    value: 'latest',
+    description: `Production-ready release (${nextPackageJson['latest'].version})`,
+  })
+
+  const shortcutVersion = process.argv[2]?.replace('@', '')
+  if (['canary', 'rc', 'latest'].includes(shortcutVersion)) {
+    // Shortcut
+    targetVersionSpecifier = shortcutVersion as VersionSpecifier
+  } else {
+    const initialVersionSpecifierIdx = await processCurrentVersion()
+
+    const response: Response = await prompts({
+      type: 'select',
+      name: 'version',
+      message: 'What Next.js version do you want to upgrade to?',
+      choices: choices,
+      initial: initialVersionSpecifierIdx,
```

Comment:
If `showRc` is `false`, and the user is on `latest`, `processCurrentVersion` will probably return index 2, which will be wrong?